core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java

```diff
@@ -449,6 +457,40 @@ public void recordSubqueries(Node node, ExpressionAnalysis expressionAnalysis)
         return unmodifiableList(quantifiedComparisonSubqueries.get(NodeRef.of(node)));
     }
 
+    public void addWindowDefinition(QuerySpecification query, String name, ResolvedWindow window)
+    {
+        if (!windowDefinitions.containsKey(NodeRef.of(query))) {
+            windowDefinitions.put(NodeRef.of(query), new LinkedHashMap<>());
+        }
+        windowDefinitions.get(NodeRef.of(query)).put(name, window);
+    }
+
+    public ResolvedWindow getWindowDefinition(QuerySpecification query, String name)
+    {
+        if (!windowDefinitions.containsKey(NodeRef.of(query))) {
+            return null;
+        }
+        return windowDefinitions.get(NodeRef.of(query)).get(name);
```

Comment:
If this method is called and the result is expected (i.e. not null is expected), the call is always followed by a check.
I didn't inline the check, because there is another use of this method which actually _should_ return null. This is when I need to assert that there's no duplicate name of window definition.